File: LeetCode/3016. Minimum Number of Pushes to Type Word II/3016.java
class Solution {

    // Function to count the frequency of each character in the word
    private int[] countFrequency(String word) {
        int[] frequency = new int[26];
        for (char c : word.toCharArray()) {
            frequency[c - 'a']++;
        }
        return frequency;
    }

    // Function to convert frequency array into a sorted list of frequencies in
    // descending order
    private List<Integer> getSortedFrequencies(int[] frequency) {
        List<Integer> frequencies = new ArrayList<>();
        for (int freq : frequency) {
            if (freq > 0) {
                frequencies.add(freq);
            }
        }
        frequencies.sort(Collections.reverseOrder());
        return frequencies;
    }

    // Function to calculate the minimum number of key presses needed
    private int calculateMinPushes(List<Integer> frequencies) {
        int totalPushes = 0;
        int keyPresses = 1;
        int keysUsed = 0;
        for (int freq : frequencies) {
            totalPushes += freq * keyPresses;
            keysUsed++;
            if (keysUsed == 8) { // Each key from 2-9 can be used once per set of 8 keys
                keysUsed = 0;
                keyPresses++;
            }
        }
        return totalPushes;
    }

    // Main function to compute minimum pushes required for the given word
    public int minimumPushes(String word) {
        int[] frequency = countFrequency(word);
        List<Integer> sortedFrequencies = getSortedFrequencies(frequency);
        return calculateMinPushes(sortedFrequencies);
    }
}

/*
 * class Solution {
 * public int minimumPushes(String word) {
 * int[] arr = new int[26];
 * for(int i=0;i<word.length();i++){
 * arr[word.charAt(i) - 'a']++;
 * }
 * Arrays.sort(arr);
 * int i=25;
 * int count = 0;
 * int start = 1;
 * int ans =0;
 * while( i >= 0 && arr[i] != 0){
 * ans+=(start * arr[i]);
 * count++;
 * if(count == 8){
 * start++;
 * count = 0;
 * }
 * i--;
 * }
 * return ans;
 * }
 * }
 */

/*
 * class Solution {
 * // Same as 3014. Minimum Number of Pushes to Type Word I
 * public int minimumPushes(String word) {
 * int ans = 0;
 * int[] count = new int[26];
 * 
 * for (final char c : word.toCharArray())
 * ++count[c - 'a'];
 * 
 * Arrays.sort(count);
 * for (int i = 0; i < 26; ++i)
 * ans += count[26 - i - 1] * (i / 8 + 1);
 * 
 * return ans;
 * }
 * }
 */

File: LeetCode/3016. Minimum Number of Pushes to Type Word II/3016.cpp
class Solution {
public:
    static int minimumPushes(string& word) {
        int freq[26]={0};
        for(char c: word) 
            freq[c-'a']++;
        sort(freq, freq+26);
    
        int sz=25, ans=0;
        for(; sz>=0 && freq[sz]!=0; sz--){
            ans+=freq[sz]*((25-sz)/8+1);    
        }
        return ans;
    }
};

/*
#include <execution>
auto f = [] {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();
class Solution {
public:
    int minimumPushes(string word) {
        array<int, 26> f {};
        for (char c : word) {
            f[c - 'a']++;
        }
        sort(execution::par_unseq, f.begin(), f.end(), greater{});
        return transform_reduce(execution::par_unseq, f.begin(), f.end(), 0, plus{}, [f = &f[0]](int const& x) { 
            int i = &x - f;
            return x * (i / 8 + 1);
        });
    }
};
*/

/*
class Solution {
public:
    static int minimumPushes(string& word) {
        int freq[26]={0};
        for(char c: word) 
            freq[c-'a']++;
        sort(freq, freq+26);
    
        int sz=25, ans=0;
        for(; sz>=0 && freq[sz]!=0; sz--){
            ans+=freq[sz]*((25-sz)/8+1);    
        }
        return ans;
    }
};
*/

/*
class Solution {
public:
    void print(auto& c){
        for(int x: c) cout<<x<<", "; 
        cout<<endl;
    }
    int minimumPushes(string& word) {
        int freq[26]={0};
    //    memset(freq, 0, sizeof(freq));
        for(char c: word) 
            freq[c-'a']++;
        sort(freq, freq+26, greater<int>());
    //    print(freq);
        int sz=0, push=1, ans=0;
        for(; sz<26 && freq[sz]!=0; sz++){
            if (sz>=8 && sz%8==0) push++;
            ans+=freq[sz]*push;    
        }
        return ans;
    }
};

auto init = []()
{ 
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 'c';
}();
*/